File: Orchestrator/src/ElizaAgent.js
const { BaseAgent } = require('./BaseAgent');
const { ConversationMemory, EmotionalMemory } = require('./memory');
const { PatternMatcher } = require('./patterns');

class ElizaAgent extends BaseAgent {
  constructor(config = {}) {
    super(config);
    
    // Initialize personality
    this.personality = {
      name: config.name || 'Eliza',
      role: config.role || 'therapist',
      style: config.style || 'empathetic'
    };

    // Initialize response style
    this.responseStyle = {
      reflectionLevel: config.reflectionLevel || 0.7,
      empathyLevel: config.empathyLevel || 0.8
    };

    // Initialize components
    this.initializeMemory(config);
    this.patternMatcher = new PatternMatcher();
  }

  initializeMemory(config) {
    const memoryType = config.memoryType || 'emotional';
    const windowSize = config.contextWindow || 5;

    if (memoryType === 'emotional') {
      this.memory = new EmotionalMemory(windowSize);
    } else {
      this.memory = new ConversationMemory(windowSize);
    }
  }

  async execute(input) {
    try {
      // Load conversation history
      const history = await this.memory.loadMemoryVariables();
      
      // Generate response based on input and history
      const response = await this.generateResponse(input, history);
      
      // Save context
      await this.memory.saveContext(
        { input },
        { output: response }
      );

      return {
        content: response.content,
        metadata: {
          type: response.type,
          emotionalState: this.memory.getEmotionalSummary?.() || null,
          personality: this.personality
        }
      };

    } catch (error) {
      console.error('Error in ElizaAgent execution:', error);
      throw error;
    }
  }

  async generateResponse(input, history) {
    // Get pattern-based response
    const patternResponse = this.patternMatcher.getResponse(
      input,
      this.responseStyle
    );

    // Format final response
    const content = this.formatResponse(patternResponse);
    
    return {
      content,
      type: patternResponse.type
    };
  }

  formatResponse(response) {
    // Add personality-specific prefixes based on role and style
    const prefixes = {
      therapist: ['I see.', 'Hmm.', 'Let\'s explore that.'],
      counselor: ['From what you\'re saying,', 'It sounds like,', 'I\'m hearing that,'],
      friend: ['You know,', 'I feel that,', 'Maybe,']
    };

    const rolePrefix = prefixes[this.personality.role] || prefixes.therapist;
    const prefix = Math.random() < 0.3 ? 
      rolePrefix[Math.floor(Math.random() * rolePrefix.length)] + ' ' :
      '';

    return prefix + response.response;
  }

  async cleanup() {
    if (this.memory) {
      await this.memory.clear();
    }
  }
}

module.exports = {
  ElizaAgent
}; 

File: python-llm-service/src/agents/eliza_agent.py
from typing import Dict, Any, List
from .base_agent import BaseAgent
from .memory import ConversationMemory, EmotionalMemory
from .pattern_matcher import PatternMatcher

class ElizaAgent(BaseAgent):
    def __init__(self, config: Dict[str, Any]):
        super().__init__(config)
        
        # Initialize personality attributes
        self.name = config.get('therapistName', 'Eliza')
        self.role = config.get('therapistRole', 'Rogerian')
        self.style = config.get('conversationStyle', 'Empathetic')
        
        # Initialize response style attributes
        self.reflection_level = config.get('reflectionLevel', 7)
        self.empathy_level = config.get('empathyLevel', 8)
        
        # Initialize memory and pattern matcher
        self.memory = self.initializeMemory(config)
        self.pattern_matcher = PatternMatcher()
        
        # Initialize social platforms
        self.social_platforms = self.initializeSocialPlatforms(config)
        
    def initializeMemory(self, config: Dict[str, Any]) -> Any:
        """Initialize the appropriate memory type based on configuration."""
        memory_type = config.get('memoryType', 'emotional')
        context_window = config.get('contextWindow', 5)
        
        if memory_type == 'emotional':
            return EmotionalMemory(window_size=context_window)
        else:
            return ConversationMemory(window_size=context_window)
            
    def initializeSocialPlatforms(self, config: Dict[str, Any]) -> Dict[str, Any]:
        """Initialize configured social platform integrations."""
        platforms = {}
        
        if 'twitter' in config.get('platforms', {}):
            from ..platforms import TwitterPlatform
            platforms['twitter'] = TwitterPlatform(config['platforms']['twitter'])
            
        if 'discord' in config.get('platforms', {}):
            from ..platforms import DiscordPlatform
            platforms['discord'] = DiscordPlatform(config['platforms']['discord'])
            
        return platforms
            
    async def execute(self, input: str) -> Dict[str, Any]:
        """Execute the Eliza agent with the given input."""
        try:
            # Load conversation history
            history = await self.memory.loadMemoryVariables({})
            
            # Generate response based on input and history
            response = self.generateResponse(input, history)
            
            # Post to configured social platforms if enabled
            social_responses = {}
            for platform_name, platform in self.social_platforms.items():
                try:
                    result = await platform.post(response['content'])
                    social_responses[platform_name] = result
                except Exception as e:
                    social_responses[platform_name] = {
                        'error': str(e),
                        'success': False
                    }
            
            # Save context
            await self.memory.saveContext(
                {"input": input},
                {"output": response}
            )
            
            return {
                "content": response['content'],
                "metadata": {
                    "role": self.role,
                    "style": self.style,
                    "reflection_level": self.reflection_level,
                    "empathy_level": self.empathy_level,
                    "social_responses": social_responses
                }
            }
            
        except Exception as e:
            return {
                "error": str(e),
                "content": "I'm having trouble processing that. Could you rephrase it?"
            }
            
    def generateResponse(self, input: str, history: Dict[str, Any]) -> Dict[str, Any]:
        """Generate a response based on input and conversation history."""
        # Get pattern-based response
        response = self.pattern_matcher.getResponse(
            input,
            {
                "reflection_level": self.reflection_level,
                "empathy_level": self.empathy_level
            }
        )
        
        # Format response based on personality
        return {
            'content': self.formatResponse(response),
            'type': response.get('type', 'default')
        }
        
    def formatResponse(self, response: str) -> str:
        """Format the response based on the agent's personality."""
        prefix = ""
        
        if self.role == "Rogerian":
            prefix = "I hear that "
        elif self.role == "Cognitive-Behavioral":
            prefix = "Let's examine "
        elif self.role == "Psychodynamic":
            prefix = "It seems that "
        elif self.role == "Humanistic":
            prefix = "I understand "
            
        return f"{prefix}{response}"
        
    async def cleanup(self):
        """Clean up resources."""
        if self.memory:
            await self.memory.clear() 